File: protocol/bridge/common/src/main/java/ink/reactor/protocol/bridge/common/handler/MutablePacketHandlerStorage.java
package ink.reactor.protocol.bridge.common.handler;

import ink.reactor.protocol.api.connection.PlayerConnection;
import ink.reactor.protocol.api.buffer.reader.ReaderBuffer;
import ink.reactor.protocol.api.packet.PacketHandler;

public interface MutablePacketHandlerStorage {
    void callHandlers(final PlayerConnection connection, final ReaderBuffer readerBuffer);

    MutablePacketHandlerStorage add(final PacketHandler handler);
    MutablePacketHandlerStorage remove(final PacketHandler handler);

    MutablePacketHandlerStorage add(final PacketHandler... handlers);
    MutablePacketHandlerStorage remove(final PacketHandler... handlers);
}

File: networking/internal/src/main/kotlin/ink/reactor/network/internal/packet/storage/MutablePacketHandlerStorage.kt
package ink.reactor.network.internal.packet.storage

import ink.reactor.network.api.packet.Packet
import ink.reactor.network.api.packet.handler.PacketHandler
import ink.reactor.network.api.player.PlayerConnection

internal interface MutablePacketHandlerStorage {
    fun callHandlers(connection: PlayerConnection, packet: Packet)

    fun add(handler: PacketHandler): MutablePacketHandlerStorage
    fun remove(handler: PacketHandler): MutablePacketHandlerStorage?

    fun add(vararg handlers: PacketHandler): MutablePacketHandlerStorage
    fun remove(vararg handlers: PacketHandler): MutablePacketHandlerStorage?
}
